odoo/cli/populate.py

```diff
@@ -1,131 +1,68 @@
-                else:
-                    profiling_context = nullcontext()
-
-                if commit:
-                    commit_context = nullcontext()
-                else:
-                    commit_context = patch('odoo.sql_db.Cursor.commit')
-
-                _logger.info('Populating database for model %s', model._name)
-                t0 = time.time()
-
-                with profiling_context, commit_context:
-                    registry.populated_models[model._name] = model._populate(size).ids
-
-                    if not registry.populated_models[model._name]:
-                        # Do not create ir.profile records
-                        # for models without any population factories
-                        profiling_context.db = False
-
-                    # force the flush to make sure population time still
-                    # considers flushing all values to database
-                    env.flush_all()
-
-                if commit:
-                    env.cr.commit()
-
-                model_time = time.time() - t0
-                if model_time > 1:
-                    _logger.info('Populated database for model %s (total: %fs) (average: %fms per record)',
-                                 model._name, model_time, model_time / len(registry.populated_models[model._name]) * 1000)
-        except:
-            _logger.exception('Something went wrong populating database')
-        finally:
-            if not commit:
-                env.cr.rollback()
-            populated_models = registry.populated_models
-            del registry.populated_models
-
-        return populated_models
-
-    @classmethod
-    def _get_ordered_models(cls, env, model_patterns=False):
-        _logger.info('Computing model order')
-        processed = set()
-        ordered_models = []
-        visited = set()
-        def add_model(model):
-            if model not in processed:
-                if model in visited:
-                    raise ValueError('Cyclic dependency detected for %s' % model)
-                visited.add(model)
-                for dep in model._populate_dependencies:
-                    add_model(env[dep])
-                ordered_models.append(model)
-                processed.add(model)
-        for model in env.values():
-            if model_patterns and not any(fnmatch.fnmatch(model._name, match) for match in model_patterns):
-                continue
-            if model._transient or model._abstract:
-                continue
-            ir_model = env['ir.model'].search([('model', '=', model._name)])
-            if not model_patterns and all(module.startswith('test_') for module in ir_model.modules.split(',')):
-                continue
-            add_model(model)
-
-        return ordered_models
+            models = [m for model_name in model_names if (m := env.get(model_name)) is not None and not (m._transient or m._abstract)]
+            factors = dict(zip(models, factors))
```

Comment:
`factors` should be the same length as `model_names`, but if some models are not in `env`, then `models` no longer corresponds to `factors`, and the dict `factors` no longer makes any sense.  This is safer are simply more readable:
```suggestion
            models = []
            model_factors = {}
            for model_name, factor in zip(model_names, factors):
                if (model := env.get(model_name)) is not None and not (model._transient or model._abstract):
                    models.append(model)
                    model_factors[model] = factor
```
By the way, keeping models and factors separate just makes everything more complex everywhere.
Why don't you use a list of pairs `(model, factor)` instead?  This would simplify `populate_models`, too.

Reply:
> Why don't you use a list of pairs (model, factor) instead? This would simplify populate_models, too.

I guess it would've worked nicely, but there is a case (when copying a Many2one) where we need the factor of the `comodel`, and not of the `model` (this can happen when both model and it's dependent model are both specified on the CLI, but with different `factors`). Due to this, I think it's easier to pass the whole dict of `factors` around, so during population of one model we may reference the factor of another model.